### ISSUE TITLE
probable memory exhaustion in travis builds.

### ISSUE THREAD

**Issue by kevingranade** [CleverRaven/Cataclysm-DDA#21420]
### Expected Behavior
Travis builds run to completion, failing only if the build or test process encounters an issue.

### Actual behavior
Frequently travis builds terminate with a message like:
```
Starting the actual test at Fri Jul 14 08:57:38 2017

/home/travis/.travis/job_stages: line 54:  9206 Killed                  tests/cata_test

The command "CXX="ccache $COMPILER" make -j3 RELEASE=1 && tests/cata_test" exited with 137.
```

### Steps to reproduce the behavior
Monitor travis builds, its unfortunately sporadic.

The following resources
https://www.google.com/url?q=https://docs.travis-ci.com/user/common-build-problems/&sa=U&ved=0ahUKEwjln-qsmonVAhUK6IMKHTG6B4oQFggLMAA&usg=AFQjCNGaAfTuJZSQFySfiCo2RXxPwGB8QA
https://github.com/travis-ci/travis-ci/issues/4192
https://stackoverflow.com/questions/27377615/why-does-travis-ci-kill-the-process-for-my-script

Indicate that it's probably memory exhaustion, and the surrounding context indicates it's the unit test executable that's responsible.

### Potential fixes
The travis issue above indicates that
```
ulimit -s 1082768 # Require a bit of more memory. 
```
May provide a workaround.

On the dev box side, it may be reproduceable by a similar ulimit statement to *reduce* available stack size for the unit test executable to match the setting on travis.
This would be followed up by memory profiling, it is likely we have a memory leak in some of our tests.

**Comment by DangerNoodle** [comment]
Out of curiosity, how many different commit checking systems have been used in the past year? I recall Jenkins and the two in current use, along with at least one used in between these.

**Comment by kevingranade** [comment]
Just two, jenkins hosted on a series of different servers and travis.
Jenkins doesn't have the sandboxing necessary to run unit tests, which is why I set up a travis account.

**Comment by DangerNoodle** [comment]
I see. Still unfortunate luck with commit testers, it would appear.

**Comment by kevingranade** [comment]
It's not luck, most likely we're going over their memory limit, its
something *we* need to fix.

**Comment by kevingranade** [comment]
```
$ /usr/bin/time -v tests/cata_test 
Loading content (dda)
Finalizing and verifying
Starting the actual test at Wed Jul 19 03:17:59 2017
...
Ended test at Wed Jul 19 03:23:31 2017
The test took 331.319 seconds
Command exited with non-zero status 2
        Command being timed: "tests/cata_test"
        User time (seconds): 333.30
        System time (seconds): 2.32
        Percent of CPU this job got: 98%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 5:41.92
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 13969356
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 150
        Minor (reclaiming a frame) page faults: 127228
        Voluntary context switches: 1227
        Involuntary context switches: 736
        Swaps: 0
        File system inputs: 62960
        File system outputs: 2304
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 2
```
tl;dr: Maximum resident set size (kbytes): 13969356
that's 13GB

So yea, we have a memory leak or something >_<

**Comment by BrettDong** [comment]
I noticed that the memory usage was steadily increasing at a constant rate on my machine, about 4 MiB per second.

**Comment by BrettDong** [comment]
Uploaded the valgrind log to Google Drive: https://drive.google.com/open?id=0B6ylQiM0nHdsVTlPcVlKOGhleGFqV1RFZF9PQ3JES0hqYmdn

**Comment by BrettDong** [comment]
Probably related to `load_mapgen_function`

**Comment by BrettDong** [comment]
After disabling two test cases, the memory leak problem was gone and the unit test finished in 68 seconds.
```diff
diff --git a/tests/monster_test.cpp b/tests/monster_test.cpp
index c3d389c579..9d804894c3 100644
--- a/tests/monster_test.cpp
+++ b/tests/monster_test.cpp
@@ -332,14 +332,14 @@ TEST_CASE("write_slope_to_speed_map_square", "[.]") {
 
 // Characterization test for monster movement speed.
 // It's not necessarally the one true speed for monsters, we just want notice if it changes.
-TEST_CASE("monster_speed_square", "[speed]") {
+TEST_CASE("monster_speed_square", "[.][speed]") {
     clear_map();
     get_options().get_option( "CIRCLEDIST" ).setValue("false");
     trigdist = false;
     monster_check();
 }
 
-TEST_CASE("monster_speed_trig", "[speed]") {
+TEST_CASE("monster_speed_trig", "[.][speed]") {
     clear_map();
     get_options().get_option( "CIRCLEDIST" ).setValue("true");
     trigdist = true;
```

**Comment by BrettDong** [comment]
`/usr/bin/time` statics with two tests shown above disabled:
```
Loading content (dda)
Finalizing and verifying
Starting the actual test at Wed Jul 19 15:37:13 2017
===============================================================================
All tests passed (1305716 assertions in 38 test cases)

Ended test at Wed Jul 19 15:38:21 2017
The test took 68.143 seconds
	Command being timed: "tests/cata_test"
	User time (seconds): 75.57
	System time (seconds): 0.27
	Percent of CPU this job got: 95%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 1:19.55
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 128200
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 77
	Minor (reclaiming a frame) page faults: 28790
	Voluntary context switches: 962
	Involuntary context switches: 14327
	Swaps: 0
	File system inputs: 38912
	File system outputs: 800
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

**Comment by BrettDong** [comment]
TL;DR: The program seems to be stuck in generating map, triggered by `can_catch_player` function in `monster_speed_*` test cases. But the problem isn't 100% reproducible. `tests/cata_test` sometimes just finishes without trouble.

GDB log
```
(gdb) bt
#0  std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, options_manager::cOpt>, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, options_manager::cOpt> >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::count (
    this=this@entry=0x1800040 <get_options()::single_instance+128>, __k="CLASSIC_ZOMBIES") at /usr/local/gcc-7.1.0/include/c++/7.1.0/bits/hashtable.h:1477
#1  0x0000000000ecd804 in std::__cxx1998::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, options_manager::cOpt, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, options_manager::cOpt> > >::count (__x="CLASSIC_ZOMBIES", this=0x1800040 <get_options()::single_instance+128>)
    at /usr/local/gcc-7.1.0/include/c++/7.1.0/bits/unordered_map.h:929
#2  options_manager::get_option (this=0x17fffc0 <get_options()::single_instance>, name="CLASSIC_ZOMBIES") at src/options.cpp:2055
#3  0x0000000000ea06ae in get_option<bool> (name="CLASSIC_ZOMBIES") at src/options.h:218
#4  MonsterGroupManager::GetResultFromGroup (group_name=..., quantity=quantity@entry=0x7fffffff5c2c, turn=turn@entry=-1) at src/mongroup.cpp:95
#5  0x0000000000bbdfa5 in map::place_spawns (this=this@entry=0x7fffffffbcf0, group=..., chance=chance@entry=2, x1=x1@entry=0, y1=y1@entry=0, x2=x2@entry=23, y2=<optimized out>, 
    density=density@entry=0.0399999991) at src/mapgen.cpp:8501
#6  0x0000000000baaf1f in mapgen_road (m=<optimized out>, terrain_type=..., dat=..., turn=<optimized out>, density=<optimized out>) at src/mapgen_functions.cpp:1304
#7  0x0000000000bb848b in mapgen_function_builtin::generate (this=<optimized out>, m=<optimized out>, o=..., mgd=..., i=<optimized out>, d=<optimized out>) at src/mapgen.cpp:217
#8  0x0000000000be0ee0 in map::draw_map (this=this@entry=0x7fffffffbcf0, terrain_type=..., t_north=..., t_east=..., t_south=..., t_west=..., t_neast=..., t_seast=..., t_swest=..., t_nwest=..., 
    t_above=..., turn=0, density=density@entry=0.0399999991, zlevel=0, rsettings=0x7ffff6b3e210) at src/mapgen.cpp:1856
#9  0x0000000000c16274 in map::generate (this=this@entry=0x7fffffffbcf0, x=x@entry=14, y=y@entry=4, z=z@entry=0, turn=0) at src/mapgen.cpp:161
#10 0x000000000075a367 in map::loadn (this=this@entry=0x1873560, gridx=gridx@entry=10, gridy=gridy@entry=3, gridz=gridz@entry=0, update_vehicles=update_vehicles@entry=true) at src/map.cpp:6700
#11 0x000000000075b402 in map::shift (this=0x1873560, sx=sx@entry=1, sy=sy@entry=1) at src/map.cpp:6496
#12 0x0000000000806369 in game::update_map (this=this@entry=0x18730e0, x=@0x7fffffffc2ac: 60, y=@0x7fffffffc2a8: 60) at src/game.cpp:12841
#13 0x00000000008068a8 in game::update_map (this=0x18730e0, p=...) at src/game.cpp:12809
#14 0x00000000004ebcbb in can_catch_player (monster_type="mon_zombie", direction_of_flight=...) at monster_test.cpp:141
#15 0x00000000004ef9da in monster_check () at monster_test.cpp:311
#16 0x00000000004f0c32 in ____C_A_T_C_H____T_E_S_T____335 () at monster_test.cpp:339
#17 0x0000000000426de5 in Catch::FreeFunctionTestCase::invoke (this=<optimized out>) at catch/catch.hpp:6591
#18 0x0000000000417cf7 in Catch::TestCase::invoke (this=<optimized out>) at catch/catch.hpp:7528
#19 0x000000000044d05e in Catch::RunContext::invokeActiveTestCase (this=0x7fffffffd9e0) at catch/catch.hpp:6167
#20 Catch::RunContext::runCurrentTest (redirectedCerr="", redirectedCout="", this=<optimized out>) at catch/catch.hpp:6138
#21 Catch::RunContext::runTest (this=this@entry=0x7fffffffd9e0, testCase=...) at catch/catch.hpp:5958
#22 0x00000000004207a4 in Catch::runTests (config=...) at catch/catch.hpp:6306
#23 0x000000000045de3d in Catch::Session::run (this=this@entry=0x7fffffffded0) at catch/catch.hpp:6414
#24 0x0000000000426a3f in main (argc=<optimized out>, argv=<optimized out>) at test_main.cpp:153
(gdb) b monster_test.cpp:141
Breakpoint 1 at 0x4ebcac: file monster_test.cpp, line 141.
(gdb) ignore 1 10
Will ignore next 10 crossings of breakpoint 1.
(gdb) c
Continuing.
^C
Program received signal SIGINT, Interrupt.
0x0000000000dd7e81 in vehicle::part_with_feature (this=this@entry=0x5685730, part=part@entry=68, flag=flag@entry=VPFLAG_OBSTACLE, unbroken=unbroken@entry=true) at src/vehicle.cpp:2069
2069	{
(gdb) bt
#0  0x0000000000dd7e81 in vehicle::part_with_feature (this=this@entry=0x5685730, part=part@entry=68, flag=flag@entry=VPFLAG_OBSTACLE, unbroken=unbroken@entry=true) at src/vehicle.cpp:2069
#1  0x0000000000dd8c33 in vehicle::obstacle_at_part (this=this@entry=0x5685730, p=p@entry=68) at src/vehicle.cpp:5888
#2  0x000000000073a2fb in map::move_cost_internal (this=this@entry=0x7fffffffbcf0, furniture=..., terrain=..., veh=0x5685730, vpart=68) at src/map.cpp:1836
#3  0x0000000000746cff in map::move_cost (this=this@entry=0x7fffffffbcf0, p=..., ignored_vehicle=ignored_vehicle@entry=0x0) at src/map.cpp:1911
#4  0x0000000000746d9e in map::passable (this=this@entry=0x7fffffffbcf0, p=...) at src/map.cpp:1921
#5  0x0000000000746dd7 in map::passable (this=this@entry=0x7fffffffbcf0, x=x@entry=11, y=y@entry=16) at src/map.cpp:1869
#6  0x0000000000746de5 in map::impassable (this=this@entry=0x7fffffffbcf0, x=x@entry=11, y=y@entry=16) at src/map.cpp:1864
#7  0x0000000000bbdf84 in map::place_spawns (this=this@entry=0x7fffffffbcf0, group=..., chance=chance@entry=2, x1=x1@entry=0, y1=y1@entry=0, x2=x2@entry=23, y2=<optimized out>, 
    density=density@entry=0.0399999991) at src/mapgen.cpp:8498
#8  0x0000000000baaf1f in mapgen_road (m=<optimized out>, terrain_type=..., dat=..., turn=<optimized out>, density=<optimized out>) at src/mapgen_functions.cpp:1304
#9  0x0000000000bb848b in mapgen_function_builtin::generate (this=<optimized out>, m=<optimized out>, o=..., mgd=..., i=<optimized out>, d=<optimized out>) at src/mapgen.cpp:217
#10 0x0000000000be0ee0 in map::draw_map (this=this@entry=0x7fffffffbcf0, terrain_type=..., t_north=..., t_east=..., t_south=..., t_west=..., t_neast=..., t_seast=..., t_swest=..., t_nwest=..., 
    t_above=..., turn=0, density=density@entry=0.0399999991, zlevel=0, rsettings=0x7ffff6b3e210) at src/mapgen.cpp:1856
#11 0x0000000000c16274 in map::generate (this=this@entry=0x7fffffffbcf0, x=x@entry=14, y=y@entry=4, z=z@entry=0, turn=0) at src/mapgen.cpp:161
#12 0x000000000075a367 in map::loadn (this=this@entry=0x1873560, gridx=gridx@entry=10, gridy=gridy@entry=3, gridz=gridz@entry=0, update_vehicles=update_vehicles@entry=true) at src/map.cpp:6700
#13 0x000000000075b402 in map::shift (this=0x1873560, sx=sx@entry=1, sy=sy@entry=1) at src/map.cpp:6496
#14 0x0000000000806369 in game::update_map (this=this@entry=0x18730e0, x=@0x7fffffffc2ac: 60, y=@0x7fffffffc2a8: 60) at src/game.cpp:12841
#15 0x00000000008068a8 in game::update_map (this=0x18730e0, p=...) at src/game.cpp:12809
#16 0x00000000004ebcbb in can_catch_player (monster_type="mon_zombie", direction_of_flight=...) at monster_test.cpp:141
#17 0x00000000004ef9da in monster_check () at monster_test.cpp:311
#18 0x00000000004f0c32 in ____C_A_T_C_H____T_E_S_T____335 () at monster_test.cpp:339
#19 0x0000000000426de5 in Catch::FreeFunctionTestCase::invoke (this=<optimized out>) at catch/catch.hpp:6591
#20 0x0000000000417cf7 in Catch::TestCase::invoke (this=<optimized out>) at catch/catch.hpp:7528
#21 0x000000000044d05e in Catch::RunContext::invokeActiveTestCase (this=0x7fffffffd9e0) at catch/catch.hpp:6167
#22 Catch::RunContext::runCurrentTest (redirectedCerr="", redirectedCout="", this=<optimized out>) at catch/catch.hpp:6138
#23 Catch::RunContext::runTest (this=this@entry=0x7fffffffd9e0, testCase=...) at catch/catch.hpp:5958
#24 0x00000000004207a4 in Catch::runTests (config=...) at catch/catch.hpp:6306
#25 0x000000000045de3d in Catch::Session::run (this=this@entry=0x7fffffffded0) at catch/catch.hpp:6414
#26 0x0000000000426a3f in main (argc=<optimized out>, argv=<optimized out>) at test_main.cpp:153
(gdb) info b
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   0x00000000004ebcac in can_catch_player(std::__cxx11::string const&, tripoint const&) at monster_test.cpp:141
	ignore next 10 hits
```

**Comment by Coolthulhu** [comment]
The tests do move the player around and then re-center the map.
It would probably be somewhat less robust, but significantly cleaner if the player always stayed in the center, but monster's position was projected to the edge of the map instead.

**Comment by kevingranade** [comment]
Very odd, I just tried it out with just that test (monster_speed_square), could see the memory growth proceeding for many minutes, but when I gave up and killed the process it was still in unit test initialization:
```
[kevingranade@buildlaptop Cataclysm-DDA]$ valgrind tests/cata_test monster_speed_square
==705== Memcheck, a memory error detector
==705== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==705== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==705== Command: tests/cata_test monster_speed_square
==705== 
Loading content (dda)
Finalizing and verifying
==705== Warning: set address range perms: large range [0x66e4c040, 0x80e4c040) (undefined)
==705== Warning: set address range perms: large range [0x80e4d040, 0x9ae4d040) (undefined)
==705== Warning: set address range perms: large range [0x2d205040, 0x47205040) (undefined)
==705== Warning: set address range perms: large range [0x9ae4e040, 0xb4e4e040) (undefined)
^C==705== 
==705== Process terminating with default action of signal 2 (SIGINT)
==705==    at 0x515C290: get_safe_base_mutex (debug.cc:59)
==705==    by 0x515C290: __gnu_debug::_Safe_sequence_base::_M_get_mutex() (debug.cc:297)
==705==    by 0x515C6AE: __gnu_debug::_Safe_sequence_base::_M_detach(__gnu_debug::_Safe_iterator_base*) (debug.cc:324)
==705==    by 0x515C863: __gnu_debug::_Safe_iterator_base::_M_detach() (debug.cc:375)
==705==    by 0x102D1DF: ~_Safe_iterator_base (safe_base.h:100)
==705==    by 0x102D1DF: ~_Safe_iterator (safe_iterator.h:86)
==705==    by 0x102D1DF: MonsterGroupManager::monster_is_blacklisted(string_id<mtype> const&) (mongroup.cpp:272)
==705==    by 0x6EB9B3: map::add_spawn(string_id<mtype> const&, int, int, int, bool, int, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (mapgen.cpp:8650)
==705==    by 0x6EBF7A: map::place_spawns(string_id<MonsterGroup> const&, int, int, int, int, int, float) (mapgen.cpp:8503)
==705==    by 0x6FC972: MapExtras::mx_helicopter(map&, tripoint const&) (mapgen.cpp:10437)
==705==    by 0x744CC4: map::generate(int, int, int, int) (mapgen.cpp:173)
==705==    by 0x97692A: map::loadn(int, int, int, bool) (map.cpp:6700)
==705==    by 0x9781A7: map::loadn(int, int, bool) (map.cpp:6626)
==705==    by 0x9785A6: map::load(int, int, int, bool) (map.cpp:6424)
==705==    by 0x4248EB: init_global_game_state(std::__debug::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (test_main.cpp:93)
==705== 
==705== HEAP SUMMARY:
==705==     in use at exit: 1,856,911,707 bytes in 304,700 blocks
==705==   total heap usage: 27,538,711 allocs, 27,234,011 frees, 4,815,714,223 bytes allocated
==705== 
==705== LEAK SUMMARY:
==705==    definitely lost: 74 bytes in 4 blocks
==705==    indirectly lost: 0 bytes in 0 blocks
==705==      possibly lost: 0 bytes in 0 blocks
==705==    still reachable: 1,856,911,633 bytes in 304,696 blocks
==705==                       of which reachable via heuristic:
==705==                         multipleinheritance: 245,552 bytes in 866 blocks
==705==         suppressed: 0 bytes in 0 blocks
==705== Rerun with --leak-check=full to see details of leaked memory
==705== 
==705== For counts of detected and suppressed errors, rerun with: -v
==705== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by kevingranade** [comment]
I checked https://travis-ci.org/CleverRaven/Cataclysm-DDA/builds and the builds started failing with 839c432
I confirmed that 46bbc50 doesn't seem to have the problem, it terminated normally under valgrind, checking 839c432 now.

**Comment by kevingranade** [comment]
839c432 did NOT exibit the problem. Re-testing (it might be probabablistic, in which case I need to retest the previous commit a few times as well).

**Comment by kevingranade** [comment]
Trying  bisect now to narrow down the original site, unfortunately it's a slow and possibly fallible reproduction case, so this might take a few tries.

**Comment by kevingranade** [comment]
Bisecton is giving me no love, it's passing on builds that failed in travis :/

Taking a closer look at 839c432, it removed some code from map::place_spawns() that would guarantee termination, and it's really suspicious that 3/3 backtraces captured during an event include map::place_spawns().  Examining MonsterGroupManager::GetResultFromGroup() does NOT fill me with confidence that it will always decrement num, in fact there are certainly scenarios with edited monster groups that will cause that behavior.

I'm going to force it to decrement num even if it fails to find a monster and see if that resolves the issue.

**Comment by kevingranade** [comment]
Thinking about symptoms, it seems like monster_speed_square just gave it more chances to trigger the bug by calling update_map() a large number of times.
